File: QuicksortSwift/QuicksortSwift/Quicksort.swift
//
//  Quicksort.swift
//  QuicksortSwift
//
//  Created by Suri on 5/23/15.
//  Copyright (c) 2015 Suri. All rights reserved.
//

import Foundation

class Quicksort {
    
    func Swap(_ A:inout [Int], first: Int, second: Int){
        let temp        = A[first];
        A[first]       = A[second];
        A[second]        = temp;
    }
    
    func partition(_ A:inout [Int], start: Int, end: Int) -> Int{
        let pivot       = A[end];
        var index       = start-1;
        
        for i in start..<end {
            if(A[i] <= pivot){
                index += 1;
                Swap(&A, first: index, second: i)
            }
            // print("index:\(index) A:\(A[end]) pivot:\(pivot)");
        }
        Swap(&A, first: index+1, second: end);
        
        return (index + 1);
    }
    func quicksort(_ v: inout [Int], start: Int, end: Int, count: inout Int) {
        if start < end {
            count += 1;
            let pivotIndex = partition(&v, start:start, end:end);
            //let pivotIndex = partition(&v, left:start, right:end)
            quicksort(&v, start: start, end: (pivotIndex - 1), count: &count);
            quicksort(&v, start: (pivotIndex + 1), end: end, count: &count);
        }
    }
}

File: QuicksortC++/QuicksortC++/QuickSort.cpp
//
//  QuickSort.m
//  QuickSort
//
//  Created by Suri on 5/23/15.
//  Copyright (c) 2015 Suri. All rights reserved.
//

void Swap (int *& A, int first , int second){
    int temp    = A[first];
    A[first]   = A[second];
    A[second]   = temp;
}

int Partition(int *& A, int p, int l){
    int pivot           = A[l];
    int i               = (p-1);
    for (int j = p; j < l; j++) {
        if(A[j] <= pivot){
            i++;
            Swap(A, i, j);
        }
    }
    Swap(A, (i+1), l);
    
    return (i + 1);
}

void QuickSortCplus(int *& A, int start, int length, int& count){
    if(start < length){
        count += 1;
        int p         = Partition(A, start, length);
        QuickSortCplus(A, start, p-1, count);
        QuickSortCplus(A, p+1, length, count);
    }
    
}


